File: exercise/isbn/preparation/values.py
"""Minimal RPC language in JSON to send data from the tests to the judge."""
import io
import json
import traceback


def encode(value):
    if value is None:
        type_ = "nothing"
        data_ = value
    elif isinstance(value, str):
        type_ = "text"
        data_ = value
    elif isinstance(value, bool):
        type_ = "boolean"
        data_ = value
    elif isinstance(value, int):
        type_ = "integer"
        data_ = value
    elif isinstance(value, float):
        type_ = "rational"
        data_ = value
    elif isinstance(value, list) or isinstance(value, tuple):
        type_ = "sequence"
        data_ = [encode(x) for x in value]
    elif isinstance(value, set):
        type_ = "set"
        data_ = [encode(x) for x in value]
    elif isinstance(value, dict):
        type_ = "map"
        data_ = {str(k): encode(v) for k, v in value.items()}
    else:
        type_ = "unknown"
        data_ = str(value)

    return {
        "data": data_,
        "type": type_
    }


def send_value(stream, value):
    """Send a value to the given stream."""
    json.dump(encode(value), stream)


def send_exception(stream, exception):
    tracer = io.StringIO()
    traceback.print_tb(exception.__traceback__, file=tracer)
    data = {
        "message": str(exception),
        "stacktrace": tracer.getvalue()
    }
    json.dump(data, stream)


# noinspection PyDefaultArgument
def send_evaluated(stream, result, expected, actual, messages):
    json.dump({
        "result": result,
        "readable_expected": expected,
        "readable_actual": actual,
        "messages": messages
    }, stream)

File: judge/src/tested/languages/javascript/templates/values.js
// Minimal RPC language in JSON to send data from the tests to the judge."""
const fs = require("fs");

function encode(value) {

    let type;

    if (typeof value === "undefined") {
        type = "unknown";
    } else if (typeof value === "boolean") {
        type = "boolean";
    } else if (typeof value === "number") {
        if (value.isInteger()) {
            type = "integer";
        } else {
            type = "rational";
        }
    } else if (typeof value === "string") {
        type = "text";
    } else if (typeof value === "bigint") {
        type = "bigint";
        value = value.toString();
    } else if (typeof value === "symbol") {
        type = "unknown";
        value = value.toString();
    } else if (typeof value === "object") {
        if (value === null) {
            type = "nothing";
        } else if (Array.isArray(value)) {
            type = "list";
            value = value.map(encode);
        } else if (value instanceof Set) {
            type = "set";
            value = Array.from(value).map(encode);
        } else if (value instanceof Map) {
            type = "set";
            value = Array
                .from(value)
                .reduce(
                    (obj, [key, value]) => (
                        // Be careful! Maps can have non-String keys; object literals can't.
                        Object.assign(obj, { [key]: encode(value) })
                    ),
                    {}
                );
        } else {
            type = "object";
        }
    } else {
        type = "unknown";
    }

    return {
        type: type,
        data: value
    };

}

// Send a value to the given stream.
function sendValue(stream, value) {
    fs.writeSync(stream, JSON.stringify(encode(value)));
}

// Send an exception to the given stream.
function sendException(stream, exception) {
    if (!exception) {
        return;
    }
    fs.writeSync(stream, JSON.stringify({
        "message": exception.toString(),
        "stacktrace": exception.stack ? exception.stack : ""
    }));
}

exports.sendValue = sendValue
exports.sendException = sendException
